Add malloc-may-fail option to goto-check

xbauch opened the pull request:
And append an `assert(false)` property to every `malloc` call.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented:
This may fix #4605.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: effdd300e96d1165003a86502e8120e48b388db3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143378007

martin-cs reviewed (commented):
Why not just change the model of what malloc does?  That seems to keep all of the functionality in one place?

https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/stdlib.c#L113

Maybe it is time to change https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/stdlib.c#L115 specifically.

xbauch commented:
> Why not just change the model of what malloc does?

I think the reasoning here is to avoid "false-positive" alarms with every `malloc` call by default, and only to be correctly reporting the potential of malloc to fail if requested.

martin-cs commented:
> I think the reasoning here is to avoid "false-positive" alarms with
> every `malloc` call by default, and only to be correctly reporting
> the potential of malloc to fail if requested.

Then have a global as part of the model that allows it to be turned on
/ off / linked to a symbolic value during analysis.  It doesn't have to
be, and shouldn't be, hard-coded to one way or the other.

hannes-steffenhagen-diffblue commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,23 +115,31 @@ inline void *malloc(__CPROVER_size_t malloc_size)
-  // make sure it's not recorded as deallocated
-  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+    __CPROVER_assert(
+      __CPROVER_malloc_may_fail == (__CPROVER_bool)0, "malloc may fail");
```
? Why is there an assert that this is false? I thought this would be more like `if (__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool()) return NULL;`

xbauch replied:
Because conditionally returning `NULL` breaks too many things down the line in CBMC.

xbauch replied:
I take it back. Returning `NULL` seems to work here just fine.

thk123 commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,23 +115,31 @@ inline void *malloc(__CPROVER_size_t malloc_size)
 
-  // make sure it's not recorded as deallocated
-  __CPROVER_deallocated=(malloc_res==__CPROVER_deallocated)?0:__CPROVER_deallocated;
+    if(__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool())
```
:no_entry_sign: Comment above needs updating

xbauch replied:
Updated.

thk123 commented on regression/cbmc/Pointer_byte_extract5/main.i:
```diff
@@ -11,13 +11,13 @@ typedef union
 {
   int Count;
-  Union List[1];
+  Union List[0];
```
:question: Why does this change? The change should go in a separate commit with a why

xbauch replied:
This was undefined behaviour (and the changed in malloc exposed that CBMC does not really support that). It now has separate commit with explanation.

tautschnig replied:
I find it rather surprising that the changes in this PR expose the lack of support for this C90-hack. What exactly was going wrong here?

xbauch replied:
Why this particular PR exposes the problem is a bit more complicated. But the problem itself can be easily reproduced, e.g. (CMBC reports the assertion as failing):
```
#include <malloc.h>
#include <assert.h>

struct Foo
{
  int i;
  char data[1];
};

int main()
{
  struct Foo* foo = malloc(sizeof(struct Foo) + sizeof(char));
  foo->data[1]='b'; // set data[1]
  assert(foo->data[1]=='b'); // check data[1] -- should succeed

  return 0;
}
```

thk123 commented on regression/cbmc/Pointer_byte_extract5/no-simplify.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE broken-smt-backend
```
:no_entry_sign: Please add an explanation / link to relevant issue as to why this is broken with smt-backend at the bottom of the desc file

xbauch replied:
Added.

thk123 commented on regression/cbmc/Pointer_byte_extract5/test.desc:
```diff
@@ -4,7 +4,7 @@ main.i
 ^SIGNAL=0$
 array\.List dynamic object upper bound in p->List\[2\]: FAILURE
-\*\* 1 of 11 failed 
+\*\* 1 of 12 failed
```
:question: Why does this change? 
:pick: Probably best to set to `\d+` since number of goals isn't what is being tested

xbauch replied:
Explained and changed.

thk123 commented on regression/cbmc/malloc-may-fail/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
```
:no_entry_sign: Please add `.desc` file _without_ the option, that checks for `VERIFICATION SUCCESSFUL`

xbauch replied:
Added.

thk123 commented on regression/goto-analyzer/constant_propagation_02/test.desc:
```diff
@@ -3,7 +3,7 @@ main.c
 ^EXIT=0$
 ^SIGNAL=0$
-^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 7, function calls: 0$
+^Simplified:  assert: 1, assume: 0, goto: 1, assigns: 8, function calls: 0$
```
:pick: Please group these changes into a separate commit with why these numbers are changing

xbauch replied:
In separate commit now.

thk123 commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1041,6 +1041,7 @@ void cbmc_parse_optionst::help()
     "\n"
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    " --malloc-may-fail            enable failures for malloc calls\n"
```
:bulb: I think it'd be better to be a little more explicit. 

```
allow malloc calls to return a null pointer
```

xbauch replied:
Changed.

thk123 commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -48,6 +48,7 @@ class optionst;
   "(document-subgoals)(outfile):(test-preprocessor)" \
   "D:I:(c89)(c99)(c11)(cpp98)(cpp03)(cpp11)" \
   "(object-bits):" \
+  "(malloc-may-fail)" \
```
:grey_question: should this option be available in other C based tools, like goto-analyzer, or does that not really make sense?

xbauch replied:
Ambivalent.

thk123 commented on src/util/config.cpp:
```diff
@@ -1093,6 +1093,11 @@ bool configt::set(const cmdlinet &cmdline)
     bv_encoding.is_object_bits_default = false;
   }
 
+  if(cmdline.isset("malloc-may-fail"))
```
:pick: ?

```suggestion
  ansi_c.malloc_may_fail = cmdline.isset("malloc-may-fail");
```

xbauch replied:
Done.

thk123 commented on regression/cbmc/malloc-may-fail/test.desc:
```diff
@@ -0,0 +1,9 @@
+^\[main.assertion.\d+\] line \d+ assertion p: FAILURE$
+^VERIFICATION FAILED$
+--
+^warning: ignoring
```
:pick: in future, we've found it helpful in test-gen to include a description explaining the purpose of the test below the line. for all tests. It helps when a test fails to understand why does it even exist. E.g

```
--
^warning: ignoring
--
Test that if --malloc-may-fail is enabled, that assert can return a null pointer
```

tautschnig commented on src/ansi-c/library/stdlib.c:
```diff
@@ -115,6 +115,9 @@ inline void *malloc(__CPROVER_size_t malloc_size)
   // and __CPROVER_allocate doesn't, but no one cares
   __CPROVER_HIDE:;
+  if(__CPROVER_malloc_may_fail && __VERIFIER_nondet___CPROVER_bool())
+    return (void *)0;
```
1) It may be more consistent to just declare a local, uninitialised variable of type `__CPROVER_bool` to avoid the side-effect of a function call.
2) The non-determinism also needs to be introduced in `calloc`.
3) `realloc` now needs to handle the case of a failing `malloc`.

xbauch replied:
Add 1. No can do: using uninitialised variables doesn't even compile with the current flag set. I can initialise it to `nondet_bool`.

xbauch replied:
2,3 are done.

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1041,6 +1041,7 @@ void cbmc_parse_optionst::help()
     "\n"
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
```
The commit message says "will be squashed" so presumably that should happen ;-)

xbauch replied:
Squashed.

tautschnig reviewed (commented):
The changes to extend support to calloc, realloc need to happen before this can be merged. The modified regression test I'd not insist as much on, but still I'd like to see clarification.

martin-cs commented:
There seems to be some duplication w.r.t. https://github.com/diffblue/cbmc/pull/5210  which is preferred?

chrisr-diffblue commented:
@martin-cs I believe the difference between this PR and #5210 is around the behaviour they are modelling. #5210 is modelling the very specific case of what happens if the user code attempts to alloc a chunk of memory that is _larger than CBMC can represent_. Where as this PR is modelling the case where "the system" (e.g. OS) cannot satisfy a malloc request because of insufficient memory availability. This is a non-deterministic situation, so this PR models the situation where malloc may non-deterministically fail.

chrisr-diffblue commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1041,6 +1041,8 @@ void cbmc_parse_optionst::help()
     "Backend options:\n"
     " --object-bits n              number of bits used for object addresses\n"
+    // NOLINTNEXTLINE(whitespace/line_length)
+    " --malloc-may-fail            allow malloc calls to return a null pointer\n"
```
Similar to #5210, should this help message move up into the "Program instrumentation options" section?

xbauch replied:
Moved to the instrumentation section + the documentation was also added.

chrisr-diffblue reviewed (commented):
Looks pretty good, but would also be super nice if you could extend the user docs at the same time ;-)

petr-bauch committed:
Add malloc-may-fail option to cbmc

and hard-code appropriate check inside our `stdlib.c`. Plus some documentation
and test.

petr-bauch committed:
Calloc may fail

using the same code as for malloc.

petr-bauch committed:
Update the constant propagation tests

More assignments due to the `__CPROVER_malloc_may_fail` instantiations.

petr-bauch committed:
Update the number of test assertions

adding one more for the malloc failing.

hannes-steffenhagen-diffblue commented on src/ansi-c/library/stdlib.c:
```diff
@@ -112,50 +137,55 @@ __CPROVER_bool __VERIFIER_nondet___CPROVER_bool();
+
+    __CPROVER_bool should_malloc_fail = __VERIFIER_nondet___CPROVER_bool();
+    __CPROVER_assert(
+      !__CPROVER_malloc_may_fail || !should_malloc_fail,
```
I don’t think it makes a lot of sense to have the assertion here in this form, this will _always_ fail the way it is written.

hannes-steffenhagen-diffblue reviewed (commented):
This should have an implementation for `realloc` as well, also I don’t think we should mix up the cases where malloc fails due to some nondeterministic condition and when it fails because we reach cprover internal limits.

martin-cs reviewed (approved):
I like this approach a lot more than the previous one.
